substrate/client/transaction-pool/src/single_state_txpool/single_state_txpool.rs

```diff
@@ -674,12 +687,15 @@ where
 			pool.validated_pool().on_block_retracted(retracted.hash);
 		}
 
-		future::join_all(
-			tree_route
-				.enacted()
-				.iter()
-				.map(|h| prune_known_txs_for_block(h, &*api, &*pool, None)),
-		)
+		// Collect extrinsics on the enacted path in a map from block hn -> extrinsics.
+		let mut extrinsics = collect_extrinsics(tree_route.enacted(), &*api).await;
+		// We keep track of everything we prune so that later we won't add
+		// transactions with those hashes from the retracted blocks.
+		let mut pruned_log = HashSet::<ExtrinsicHash<PoolApi>>::new();
+		future::join_all(tree_route.enacted().iter().map(|h| {
+			let xts = extrinsics.remove(&h.hash).unwrap_or_else(|| Vec::new());
+			prune_known_txs_for_block(h, &*api, &*pool, xts, None)
+		}))
```

Comment:
I am super paranoid about any non-trivial changes in old pool.

Here you are changing the flow. How can you ensure nothing was broken?

Maybe let `extrinsics` argument of  `prune_known_txs_for_block` be the option - and if `None` is provided, let's fetch body. Old pool will provide `None`, new will provided shared buffer. We will keep the old flow of futures exactly the same.

Reply:
Thanks for the suggestion, I agree we can keep the old pool work as before. Will change soon.